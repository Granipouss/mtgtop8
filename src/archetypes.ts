import { getPage } from './utils';

export type ShortDeck = {
  id: number;
  eventId: number;
  name: string;
};

export type Archetype = { id: number; name: string; decks: ShortDeck[] };

const parseName = ($el: Cheerio): string =>
  $el
    .eq(1)
    .text()
    .trim()
    .replace(/ decks$/, '');

const parseDeck = ($el: Cheerio): ShortDeck => {
  const regexp = /^<a href="event\?e=(\d+)&amp;d=(\d+)&amp;f=MO">(.+)<\/a>$/;
  const match = $el.find('td').eq(1).html()?.match(regexp);
  if (!match) throw new Error(`Cannot parse deck: ${$el.text()}`);

  const eventId = parseInt(match[1], 10);
  const id = parseInt(match[2], 10);
  const name = match[3].trim();

  return { eventId, id, name };
};

export const getArchetype = async (
  id: number,
  page = 1
): Promise<Archetype> => {
  const $ = await getPage(
    `https://www.mtgtop8.com/archetype?a=${id}&current_page=${page}`
  );

  const archetype: Archetype = {
    id,
    name: parseName($('.w_title')),
    decks: [],
  };

  $('table table[align=center] .hover_tr').each((_, el) => {
    archetype.decks.push(parseDeck($(el)));
  });

  return archetype;
};
