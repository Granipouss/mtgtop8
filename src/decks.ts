import { getPage } from './utils';

export type ShortArchetype = { name: string; id: number };

export type CardGroup = 'mainboard' | 'sideboard' | 'commander';

export type Card = { count: number; name: string; isCompanion?: boolean };

export type Deck = {
  id: number;
  eventId: number;
  format: string;
  name: string;
  archetype: ShortArchetype;
  player: string;
  result: string;
} & {
  [group in CardGroup]: Card[];
};

const parseFormat = ($el: Cheerio): string => {
  const match = $el.html()?.match(/(.+) <div/);
  if (!match) throw new Error(`Cannot parse format: ${$el.text()}`);
  return match[1].trim();
};

const parseArchetype = ($el: Cheerio): ShortArchetype => {
  const regexp = /^<a href="archetype\?a=(\d+)">(.+)<br>decks<\/a>$/;
  const match = $el.html()?.match(regexp);
  if (!match) throw new Error(`Cannot parse archetype: ${$el.text()}`);

  const id = parseInt(match[1], 10);
  const name = match[2].trim();

  return { id, name };
};

const parseCard = ($el: Cheerio): Card => {
  const text = $el.text();
  const match = text.match(/^(\d+) (.+)$/);
  if (!match) throw new Error(`Cannot parse card: ${text}`);

  const count = parseInt(match[1], 10);
  const name = match[2].trim();
  const isCompanion = $el.find('[id="companion_%%%"]').length > 0;

  const card: Card = { count, name };
  if (isCompanion) card.isCompanion = true;

  return card;
};

const parseGroup = ($el: Cheerio): CardGroup => {
  switch ($el.text()) {
    case 'COMMANDER':
      return 'commander';
    case 'SIDEBOARD':
      return 'sideboard';
    default:
      return 'mainboard';
  }
};

export const getDeck = async (eventId: number, id: number): Promise<Deck> => {
  const $ = await getPage(`https://www.mtgtop8.com/event?e=${eventId}&d=${id}`);

  if ($.html().includes('No event could be found.')) {
    throw new Error(`No deck ${id} for event ${eventId}`);
  }

  const deck: Deck = {
    id,
    eventId,
    name: $('table .chosen_tr :not([align=right]) .topic').text().trim(),
    player: $('table .chosen_tr [align=right] .topic').text().trim(),
    result: $('table .chosen_tr [align=center]').text().trim(),
    format: parseFormat($('table table td[align="center"]').first()),
    archetype: parseArchetype($('.Nav_link').eq(1)),
    mainboard: [],
    sideboard: [],
    commander: [],
  };

  let currentGroup: CardGroup = 'mainboard';
  $('table table table td td').each((_, el) => {
    const isCard = $('.hover_tr', el).length > 0;
    if (isCard) {
      deck[currentGroup].push(parseCard($(el)));
    } else {
      currentGroup = parseGroup($(el));
    }
  });

  return deck;
};
