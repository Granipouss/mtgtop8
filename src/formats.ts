import { getPage } from './utils';

export type Format = { name: string; code: string };

export const getFormats = async (): Promise<Format[]> => {
  const $ = await getPage(`https://www.mtgtop8.com/`);

  const formats: Format[] = [];

  $('table')
    .eq(1)
    .find('td')
    .each((_, el) => {
      const link = $(el).find('a').attr('href');
      const linkRegexp = /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/;
      const linkMatch = link?.match(linkRegexp);
      if (!linkMatch) return;
      const code = linkMatch[1].trim().toUpperCase();

      const alt = $(el).find('img').attr('id');
      if (!alt) return;
      const name = alt
        .replace(/^menu_/, '')
        .replace(/_/g, ' ')
        .toLowerCase();

      formats.push({ name, code });
    });

  $('#other_tooltip_content div').each((_, el) => {
    const link = $(el).find('a').attr('href');
    const linkRegexp = /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/;
    const linkMatch = link?.match(linkRegexp);
    if (!linkMatch) return;
    const code = linkMatch[1].trim().toUpperCase();

    const alt = $(el).find('img').attr('id');
    if (!alt) return;
    const name = alt
      .replace(/^menu_/, '')
      .replace(/_/g, ' ')
      .toLowerCase();

    formats.push({ name, code });
  });

  return formats;
};
