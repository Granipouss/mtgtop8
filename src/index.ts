export * from './formats';
export * from './decks';
export * from './archetypes';
