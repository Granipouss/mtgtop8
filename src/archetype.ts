import { getPage, url, match } from './utils';

import { ShortDeck } from './deck';

export type ShortArchetype = { id: number; name: string };

export class Archetype implements ShortArchetype {
  public static async get(id: number, page = 1): Promise<Archetype> {
    const $ = await getPage(url.archetype(id, page));
    return new Archetype($);
  }

  id: number;
  name: string;

  decks: ShortDeck[];

  private constructor($: CheerioStatic) {
    this.id = parseInt(match($('table table option').attr('value'), /archetype\?a=(\d+)&f=.+&meta=.+/)[0], 10);

    this.name = $('table table').eq(1).find('tr:first-child').text().replace('decks', '').trim();

    this.decks = $('table table[align=center] .hover_tr')
      .get()
      .map((el) => {
        const [eventString, deckString, name] = match(
          $(el).find('td').eq(1).html() || '',
          /^<a href="event\?e=(\d+)&amp;d=(\d+)&amp;f=MO">(.+)<\/a>$/
        );

        return {
          eventId: parseInt(eventString, 10),
          id: parseInt(deckString, 10),
          name: name.trim(),
        };
      });
  }
}
