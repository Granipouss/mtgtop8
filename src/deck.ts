import { getPage, url, match, ListOf, WithCount, parseDeckLink } from './utils';

import { ShortArchetype } from './archetype';
import { ShortFormat } from './format';

export type ShortDeck = { eventId: number; id: number; name: string };

export type CardGroup = 'mainboard' | 'sideboard' | 'commander';

export type Card = { name: string; isCompanion?: boolean };

export class Deck implements ShortDeck {
  public static async get(eventId: number, id: number): Promise<Deck> {
    const $ = await getPage(url.deck(eventId, id));

    // FIXME: check for deck also
    if ($.html().includes('No event could be found.')) {
      throw new Error(`No deck ${id} for event ${eventId}`);
    }

    return new Deck($);
  }

  eventId: number;
  id: number;
  name: string;

  format: ShortFormat;
  archetype: ShortArchetype;
  player: string;
  result: string;

  mainboard: ListOf<Card> = [];
  sideboard: ListOf<Card> = [];
  commander: ListOf<Card> = [];

  private constructor($: CheerioStatic) {
    const chosenLinkData = parseDeckLink($('.chosen_tr a').attr('href'));

    this.eventId = chosenLinkData.event;

    this.id = chosenLinkData.deck;

    this.name = $('table .chosen_tr :not([align=right]) .topic').text().trim();

    this.format = {
      name: match($('table table td[align="center"]').html() || '', /(.+) <div/)[0].trim(),
      code: chosenLinkData.format,
    };

    const [archetypeId, archetypeName] = match(
      $('.Nav_link').eq(1).html() || '',
      /^<a href="archetype\?a=(\d+)">(.+)<br>decks<\/a>$/
    );

    this.archetype = {
      id: parseInt(archetypeId, 10),
      name: archetypeName.trim(),
    };

    this.player = $('table .chosen_tr [align=right] .topic').text().trim();

    this.result = $('table .chosen_tr [align=center]').text().trim();

    let currentGroup: CardGroup = 'mainboard';
    $('table table table td td').each((_, el) => {
      const isCard = $('.hover_tr', el).length > 0;
      if (isCard) {
        this[currentGroup].push(parseCard($(el)));
      } else {
        currentGroup = parseGroup($(el));
      }
    });
  }
}

const parseCard = ($el: Cheerio): WithCount<Card> => {
  const text = $el.text();
  const match = text.match(/^(\d+) (.+)$/);
  if (!match) throw new Error(`Cannot parse card: ${text}`);

  const count = parseInt(match[1], 10);
  const name = match[2].trim();
  const isCompanion = $el.find('[id="companion_%%%"]').length > 0;

  const card: WithCount<Card> = { count, name };
  if (isCompanion) card.isCompanion = true;

  return card;
};

const parseGroup = ($el: Cheerio): CardGroup => {
  switch ($el.text()) {
    case 'COMMANDER':
      return 'commander';
    case 'SIDEBOARD':
      return 'sideboard';
    default:
      return 'mainboard';
  }
};
