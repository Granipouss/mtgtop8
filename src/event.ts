import { getPage, url, match, parseDeckLink, isDefined } from './utils';
import { ShortFormat } from './format';
import { ShortDeck } from './deck';

export type ShortEvent = { id: number; name: string };

type WithRanks<T> = T & { ranks: number[] };

export class Event implements ShortEvent {
  public static async get(id: number): Promise<Event> {
    const $ = await getPage(url.event(id));

    if ($.html().includes('No event could be found.')) {
      throw new Error(`No event ${id}`);
    }

    return new Event($);
  }

  id: number;
  name: string;

  format: ShortFormat;
  rank: number;
  playerCount: number;
  date: Date;
  decks: WithRanks<ShortDeck>[] = [];

  private constructor($: CheerioStatic) {
    const chosenLinkData = parseDeckLink($('.chosen_tr a').attr('href'));

    this.id = chosenLinkData.event;

    this.name = $('.w_title td:first-child').text().trim();

    this.format = {
      name: match($('table table td[align="center"]').html() || '', /(.+) <div/)[0].trim(),
      code: chosenLinkData.format,
    };

    const stars = $('img[src="/graph/star.png"]').length;
    const bigstars = $('img[src="/graph/bigstar.png"]').length;
    this.rank = 10 * bigstars + stars;

    const [playerCount, date] = match($('table table td[align="center"]').html() || '', /(\d+) players - (.+)/);

    this.playerCount = parseInt(playerCount, 10);

    this.date = new Date(`${date}Z`);

    this.decks = $('.hover_tr, .chosen_tr')
      .get()
      .map((el) => {
        try {
          const linkData = parseDeckLink($(el).find('a').first().attr('href'));
          return {
            ranks: $(el).find('div:first-child').text().split('-').map(Number),
            eventId: linkData.event,
            id: linkData.deck,
            name: $(el).find('a').first().text(),
          };
        } catch (error) {
          return;
        }
      })
      .filter(isDefined);
  }
}
