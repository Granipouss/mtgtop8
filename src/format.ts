import { getPage, match, ListOf, isDefined, url } from './utils';

import { ShortArchetype } from './archetype';
import { ShortEvent } from './event';

export type ShortFormat = { name: string; code: string };

export class Format implements ShortFormat {
  public static async get(code: string): Promise<Format> {
    const $ = await getPage(url.format(code));

    if (getTableRows($, /Metagame Breakdown/).length === 1) {
      throw new Error(`No format for code ${code}`);
    }

    return new Format($);
  }

  public static async getAll(): Promise<ShortFormat[]> {
    const $ = await getPage(url.base());
    return [...$('table').eq(1).find('td').get(), ...$('#other_tooltip_content div').get()]
      .map((el) => {
        try {
          const [code] = match($(el).find('a').attr('href'), /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/);

          const alt = $(el).find('img').attr('id');
          if (!alt) return;
          const name = alt
            .replace(/^menu_/, '')
            .replace(/_/g, ' ')
            .toLowerCase();

          return { name, code };
        } catch (error) {
          return;
        }
      })
      .filter(isDefined);
  }

  public name: string;
  public code: string;

  public archetypes: ListOf<ShortArchetype> = [];

  private constructor($: CheerioStatic) {
    this.code = match($('table table .hover_tr a').attr('href'), /archetype\?a=.+&meta=.+&f=(.+)/)[0];

    this.name = $('table table a[target=_blank]')
      .text()
      .replace(/ Construction Rules$/, '')
      .toLowerCase()
      .trim();

    this.archetypes = getTableRows($, /Metagame Breakdown/)
      .filter((el) => $(el).is('.hover_tr'))
      .map((el) => {
        const [idString] = match($(el).find('a').attr('href'), /archetype\?a=(\d+)&meta=.+&f=.+/);
        return {
          id: parseInt(idString, 10),
          name: $(el).find('a').text().trim(),
          count: parseInt($(el).find('[align=right]').first().text(), 10),
        };
      });
  }

  async getEvents(page = 1): Promise<ShortEvent[]> {
    const $ = await getPage(url.format(this.code, page));
    return getTableRows($, /^(Events \d+ to \d+|Last \d+ events)$/).map((el) => {
      const [idString] = match($(el).find('a').attr('href'), /event\?e=(\d+)&f=.+/);
      return {
        id: parseInt(idString, 10),
        name: $(el).find('a').text().trim(),
      };
    });
  }
}

const getTableRows = ($: CheerioStatic, title: RegExp): CheerioElement[] => {
  try {
    const table: CheerioElement | undefined = $('table table')
      .get()
      .find((el) => $('tr:first-child', el).text().match(title));
    return table ? $(table).find('tr').get().splice(1) : [];
  } catch (error) {
    return [];
  }
};
