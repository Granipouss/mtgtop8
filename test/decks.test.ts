import { getDeck } from '../src/decks';

describe('Decks', () => {
  it('should match snapshot for Pioneer deck', async () => {
    const data = await getDeck(26845, 409141);
    expect(data).toMatchSnapshot();
  });

  it('should match snapshot for EDH deck', async () => {
    const data = await getDeck(26139, 397720);
    expect(data).toMatchSnapshot();
  });
});
