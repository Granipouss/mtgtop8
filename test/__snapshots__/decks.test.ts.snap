// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decks should match snapshot for EDH deck 1`] = `
Object {
  "archetype": Object {
    "id": 911,
    "name": "Kinnan, Bonder Prodigy",
  },
  "commander": Array [
    Object {
      "count": 1,
      "name": "Kinnan, Bonder Prodigy",
    },
  ],
  "eventId": 26139,
  "format": "Duel Commander",
  "id": 397720,
  "mainboard": Array [
    Object {
      "count": 1,
      "name": "Botanical Sanctum",
    },
    Object {
      "count": 1,
      "name": "Breeding Pool",
    },
    Object {
      "count": 1,
      "name": "Command Tower",
    },
    Object {
      "count": 1,
      "name": "Dryad Arbor",
    },
    Object {
      "count": 1,
      "name": "Flooded Strand",
    },
    Object {
      "count": 1,
      "name": "Forest",
    },
    Object {
      "count": 1,
      "name": "Hinterland Harbor",
    },
    Object {
      "count": 1,
      "name": "Misty Rainforest",
    },
    Object {
      "count": 1,
      "name": "Oboro, Palace in the Clouds",
    },
    Object {
      "count": 1,
      "name": "Polluted Delta",
    },
    Object {
      "count": 1,
      "name": "Scalding Tarn",
    },
    Object {
      "count": 1,
      "name": "Seat of the Synod",
    },
    Object {
      "count": 8,
      "name": "Snow-Covered Forest",
    },
    Object {
      "count": 7,
      "name": "Snow-Covered Island",
    },
    Object {
      "count": 1,
      "name": "Tropical Island",
    },
    Object {
      "count": 1,
      "name": "Verdant Catacombs",
    },
    Object {
      "count": 1,
      "name": "Waterlogged Grove",
    },
    Object {
      "count": 1,
      "name": "Windswept Heath",
    },
    Object {
      "count": 1,
      "name": "Wooded Foothills",
    },
    Object {
      "count": 1,
      "name": "Yavimaya Coast",
    },
    Object {
      "count": 1,
      "name": "Birds of Paradise",
    },
    Object {
      "count": 1,
      "name": "Bloom Tender",
    },
    Object {
      "count": 1,
      "name": "Boreal Druid",
    },
    Object {
      "count": 1,
      "name": "Devoted Druid",
    },
    Object {
      "count": 1,
      "name": "Drift of Phantasms",
    },
    Object {
      "count": 1,
      "name": "Drover of the Mighty",
    },
    Object {
      "count": 1,
      "name": "Elvish Mystic",
    },
    Object {
      "count": 1,
      "name": "Eternal Witness",
    },
    Object {
      "count": 1,
      "name": "Fyndhorn Elves",
    },
    Object {
      "count": 1,
      "name": "Harabaz Druid",
    },
    Object {
      "count": 1,
      "name": "Ilysian Caryatid",
    },
    Object {
      "count": 1,
      "name": "Incubation Druid",
    },
    Object {
      "count": 1,
      "name": "Llanowar Elves",
    },
    Object {
      "count": 1,
      "name": "Manaweft Sliver",
    },
    Object {
      "count": 1,
      "name": "Maraleaf Pixie",
    },
    Object {
      "count": 1,
      "name": "Painter's Servant",
    },
    Object {
      "count": 1,
      "name": "Paradise Druid",
    },
    Object {
      "count": 1,
      "name": "Reclamation Sage",
    },
    Object {
      "count": 1,
      "name": "Scavenging Ooze",
    },
    Object {
      "count": 1,
      "name": "Spectral Sailor",
    },
    Object {
      "count": 1,
      "name": "Spellseeker",
    },
    Object {
      "count": 1,
      "name": "Sylvan Caryatid",
    },
    Object {
      "count": 1,
      "name": "Thrasios, Triton Hero",
    },
    Object {
      "count": 1,
      "name": "Tribute Mage",
    },
    Object {
      "count": 1,
      "name": "Trinket Mage",
    },
    Object {
      "count": 1,
      "name": "Trophy Mage",
    },
    Object {
      "count": 1,
      "name": "Urza, Lord High Artificer",
    },
    Object {
      "count": 1,
      "name": "Utopia Tree",
    },
    Object {
      "count": 1,
      "name": "Walking Ballista",
    },
    Object {
      "count": 1,
      "name": "Bounty of the Hunt",
    },
    Object {
      "count": 1,
      "name": "Brainstorm",
    },
    Object {
      "count": 1,
      "name": "Daze",
    },
    Object {
      "count": 1,
      "name": "Dramatic Reversal",
    },
    Object {
      "count": 1,
      "name": "Fabricate",
    },
    Object {
      "count": 1,
      "name": "Force of Vigor",
    },
    Object {
      "count": 1,
      "name": "Force of Will",
    },
    Object {
      "count": 1,
      "name": "Gifts Ungiven",
    },
    Object {
      "count": 1,
      "name": "Green Sun's Zenith",
    },
    Object {
      "count": 1,
      "name": "Intuition",
    },
    Object {
      "count": 1,
      "name": "Invigorate",
    },
    Object {
      "count": 1,
      "name": "Mental Misstep",
    },
    Object {
      "count": 1,
      "name": "Miscalculation",
    },
    Object {
      "count": 1,
      "name": "Muddle the Mixture",
    },
    Object {
      "count": 1,
      "name": "Noxious Revival",
    },
    Object {
      "count": 1,
      "name": "Once Upon a Time",
    },
    Object {
      "count": 1,
      "name": "Ponder",
    },
    Object {
      "count": 1,
      "name": "Veil of Summer",
    },
    Object {
      "count": 1,
      "name": "Worldly Tutor",
    },
    Object {
      "count": 1,
      "name": "Arcane Signet",
    },
    Object {
      "count": 1,
      "name": "Basalt Monolith",
    },
    Object {
      "count": 1,
      "name": "Birthing Pod",
    },
    Object {
      "count": 1,
      "name": "Cryptolith Rite",
    },
    Object {
      "count": 1,
      "name": "Earthcraft",
    },
    Object {
      "count": 1,
      "name": "Freed from the Real",
    },
    Object {
      "count": 1,
      "name": "Garruk Relentless",
    },
    Object {
      "count": 1,
      "name": "Grindstone",
    },
    Object {
      "count": 1,
      "name": "Isochron Scepter",
    },
    Object {
      "count": 1,
      "name": "Jace, Wielder of Mysteries",
    },
    Object {
      "count": 1,
      "name": "Leyline of Abundance",
    },
    Object {
      "count": 1,
      "name": "Mox Amber",
    },
    Object {
      "count": 1,
      "name": "Oko, Thief of Crowns",
    },
    Object {
      "count": 1,
      "name": "Paradise Mantle",
    },
    Object {
      "count": 1,
      "name": "Pemmin's Aura",
    },
    Object {
      "count": 1,
      "name": "Springleaf Drum",
    },
    Object {
      "count": 1,
      "name": "Squirrel Nest",
    },
    Object {
      "count": 1,
      "name": "Sylvan Library",
    },
  ],
  "name": "Kinnah",
  "player": "Marek Vard≈æik",
  "result": "5-8",
  "sideboard": Array [],
}
`;

exports[`Decks should match snapshot for Pioneer deck 1`] = `
Object {
  "archetype": Object {
    "id": 887,
    "name": "Fires of Invention",
  },
  "commander": Array [],
  "eventId": 26845,
  "format": "Pioneer",
  "id": 409141,
  "mainboard": Array [
    Object {
      "count": 1,
      "name": "Castle Vantress",
    },
    Object {
      "count": 2,
      "name": "Clifftop Retreat",
    },
    Object {
      "count": 1,
      "name": "Geier Reach Sanitarium",
    },
    Object {
      "count": 3,
      "name": "Glacial Fortress",
    },
    Object {
      "count": 4,
      "name": "Hallowed Fountain",
    },
    Object {
      "count": 4,
      "name": "Interplanar Beacon",
    },
    Object {
      "count": 1,
      "name": "Island",
    },
    Object {
      "count": 1,
      "name": "Mountain",
    },
    Object {
      "count": 1,
      "name": "Plains",
    },
    Object {
      "count": 4,
      "name": "Raugrin Triome",
    },
    Object {
      "count": 4,
      "name": "Sacred Foundry",
    },
    Object {
      "count": 4,
      "name": "Steam Vents",
    },
    Object {
      "count": 2,
      "name": "Sulfur Falls",
    },
    Object {
      "count": 1,
      "name": "Temple of Enlightenment",
    },
    Object {
      "count": 1,
      "name": "Temple of Epiphany",
    },
    Object {
      "count": 1,
      "name": "Temple of Triumph",
    },
    Object {
      "count": 4,
      "name": "Fae of Wishes",
    },
    Object {
      "count": 3,
      "name": "Chance for Glory",
    },
    Object {
      "count": 2,
      "name": "Day's Undoing",
    },
    Object {
      "count": 2,
      "name": "Deafening Clarion",
    },
    Object {
      "count": 3,
      "name": "Supreme Verdict",
    },
    Object {
      "count": 1,
      "name": "Detention Sphere",
    },
    Object {
      "count": 2,
      "name": "Elspeth Conquers Death",
    },
    Object {
      "count": 1,
      "name": "Elspeth, Sun's Champion",
    },
    Object {
      "count": 4,
      "name": "Fires of Invention",
    },
    Object {
      "count": 3,
      "name": "Gideon of the Trials",
    },
    Object {
      "count": 2,
      "name": "Jace, Architect of Thought",
    },
    Object {
      "count": 1,
      "name": "Narset of the Ancient Way",
    },
    Object {
      "count": 3,
      "name": "Narset, Parter of Veils",
    },
    Object {
      "count": 2,
      "name": "Oath of Chandra",
    },
    Object {
      "count": 1,
      "name": "Oath of Teferi",
    },
    Object {
      "count": 4,
      "name": "Omen of the Sea",
    },
    Object {
      "count": 2,
      "name": "Shark Typhoon",
    },
    Object {
      "count": 1,
      "name": "Teferi, Hero of Dominaria",
    },
    Object {
      "count": 4,
      "name": "Teferi, Time Raveler",
    },
  ],
  "name": "Jeskai Fires of Invention",
  "player": "yunk3r",
  "result": "1",
  "sideboard": Array [
    Object {
      "count": 1,
      "isCompanion": true,
      "name": "Yorion, Sky Nomad",
    },
    Object {
      "count": 1,
      "name": "Anger of the Gods",
    },
    Object {
      "count": 1,
      "name": "Casualties of War",
    },
    Object {
      "count": 1,
      "name": "Chandra, Awakened Inferno",
    },
    Object {
      "count": 1,
      "name": "Day's Undoing",
    },
    Object {
      "count": 2,
      "name": "Dovin's Veto",
    },
    Object {
      "count": 1,
      "name": "Gideon of the Trials",
    },
    Object {
      "count": 1,
      "name": "Narset, Parter of Veils",
    },
    Object {
      "count": 1,
      "name": "Oath of Teferi",
    },
    Object {
      "count": 2,
      "name": "Rest in Peace",
    },
    Object {
      "count": 1,
      "name": "Supreme Verdict",
    },
    Object {
      "count": 1,
      "name": "Thought Distortion",
    },
    Object {
      "count": 1,
      "name": "Unmoored Ego",
    },
  ],
}
`;
